src/framework/mpas_field_routines.F

```diff
@@ -1860,139 +1707,961 @@ end subroutine mpas_deallocate_field4d_real!}}}
+
+       implicit none
+
+       type (field1dChar), pointer :: f !< Input: Field to deallocate
+
+       call mpas_deallocate_field_target(f, keep_head_ptr=.false.)
+       nullify(f)
+
+   end subroutine mpas_deallocate_field1d_char!}}}
+
+
+!***********************************************************************
+!
+!  routine mpas_deallocate_field0D_logical_target
+!
+!> \brief   MPAS 0D int deallocation routine.
+!> \author  Doug Jacobsen, Michael G. Duda
+!> \date    04/02/13
+!> \details 
+!>  This routine deallocates a 0D logical field. If the optional argument
+!>  keep_head_ptr is .true., the field type pointed to by f will not be
+!>  deallocated -- only the contents (array, attLists, constituentNames)
+!>  will be deallocated.
+!>
+!>  The principal use for the keep_head_ptr argument is for situations in
+!>  which elemements of an array of fields need to be deallocated.
+!
+!-----------------------------------------------------------------------
+   subroutine mpas_deallocate_field0d_logical_target(f, keep_head_ptr)!{{{
+
+       implicit none
+
+       type (field0dLogical), target :: f !< Input: Field to deallocate
+       logical, intent(in), optional :: keep_head_ptr
+
+       type (field0dLogical), pointer :: f_cursor, f_next
+       logical :: local_keep_head_ptr
        integer :: threadNum
        integer :: i, iErr
 
        threadNum = mpas_threading_get_thread_num()
 
+       if (present(keep_head_ptr)) then
+           local_keep_head_ptr = keep_head_ptr
+       else
+           local_keep_head_ptr = .false.
+       end if
+
        if ( threadNum == 0 ) then
-          f_cursor => f
-          do while(associated(f_cursor))
-            if(associated(f % next)) then
-              f => f % next
-            else
-              nullify(f)
-            end if
-
-            if(associated(f_cursor % array)) then
-              deallocate(f_cursor % array)
-            end if
-
-            if ( associated(f_cursor % attLists) ) then
-               do i = 1, size(f_cursor % attLists, dim=1)
-                  call mpas_deallocate_attlist(f_cursor % attLists(i) % attList, iErr)
-               end do
-               deallocate(f_cursor % attLists)
-            end if
+           f_cursor => f
+           do while(associated(f_cursor))
+               f_next => f_cursor % next
+
+               if (associated(f_cursor % attLists)) then
+                   do i = 1, size(f_cursor % attLists, dim=1)
+                       call mpas_deallocate_attlist(f_cursor % attLists(i) % attList, iErr)
+                   end do
+                   deallocate(f_cursor % attLists)
+               end if
+
+               if (associated(f_cursor % constituentNames)) then
+                   deallocate(f_cursor % constituentNames)
+               end if
 
-            deallocate(f_cursor)
+               if (.not. local_keep_head_ptr) then
+                   deallocate(f_cursor)
+               end if
```

Comment:
Please note that IBM compiler (`xlf2008_r v.16.1.1-3`) stops on this `deallocate` with run-time error:
```
1525-109 Error encountered while attempting to deallocate a data object.  The program will stop.
```
After re-running with additional `STAT` and `ERRMSG` arguments, deallocate returns status value of `2: An invalid data object has been specified for the DEALLOCATE statement.`. It appears that the standard wants the object to be either an allocatable array or an object with `target` attribute. That's not the case here with `f_cursor` declared with `pointer` attribute.

Grepping for calls to this subroutine, all callers either 
1. call with `keep_head_ptr=.true.` and deallocate the object immediately after the call, or
2. call with `keep_head_ptr=.false.` and nullify the object right after the call.

If I comment out the `if` statement, the run proceeds with no errors. It appears that all calls need to pass in `keep_head_ptr=.true.` (or just remove this logic and the corresponding call to `deallocate` the head pointer). Would that be OK, @mgduda ?

Reply:
I'll take a closer look. I'm hesitant to remove a `DEALLOCATE` statement without understanding whether this might re-introduce a memory leak under certain circumstances. From what I recall of the 2003 standard, objects with the `POINTER` attribute implicitly have the `TARGET` attribute.

Reply:
Here is a branch that lets me build and run a sea-ice case: [diff](https://github.com/MPAS-Dev/MPAS-Model/compare/develop...amametjanov:framework/nullify-field-pointers). I can try valgrind on the case that had mem-leaks. How to configure one?

When I try `SMS.T62_oQU120.CMPASO-NYF` with `config_time_integrator = 'RK4'` the case runs into `SIGFPE: erroneous arithmetic operation`. Does an RK4 case run in E3SM?

Reply:
Sorry for not following-up on this issue for so long. I just encountered the same issue on Summit, and I'll see what I can find.

Reply:
I just opened Issue #481 to track progress on this.